src/libraries/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -881,17 +881,64 @@ public ImmutableArray<T> RemoveRange(ReadOnlySpan<T> items, IEqualityComparer<T>
                 return self.Remove(items[0], equalityComparer);
             }
 
-            var indicesToRemove = new SortedSet<int>();
-            foreach (var item in items)
+#nullable disable
+            var itemsDic = new Dictionary<T, int>(equalityComparer);
+#nullable restore
+            int nullValueCount = 0;
+            foreach (ref readonly T item in items)
             {
-                int index = -1;
-                do
+                if (item == null)
                 {
-                    index = self.IndexOf(item, index + 1, equalityComparer);
-                } while (index >= 0 && !indicesToRemove.Add(index) && index < self.Length - 1);
+                    nullValueCount++;
+                }
+                else if (itemsDic.TryGetValue(item, out int count))
```

Comment:
This is reading the dictionary twice for lookup and insertion. You might be able to combine the two using the new [`CollectionsMarshal.GetValueRefOrAddDefault`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluereforadddefault?view=net-6.0) method.

Reply:
Changed, and separated 2 RemoveChange() implementation (current lookup and `CollectionsMarshal.GetValueRefOrAddDefault`) into 2 tfm files because `CollectionsMarshal.GetValueRefOrAddDefault` is from .net 6. Correct my understand if wrong, thanks.